### PR TITLE
Issue 1117  lincomboperator adjoint

### PR DESCRIPTION
**Pull request by adler-j**

### CODE REVIEW

File: odl/operator/default_ops.py [hunk 1]
```diff
@@ -171,53 +171,157 @@ def __str__(self):
...
+    def __init__(self, space, *coeffs):
         """Initialize a new instance.
 
         Parameters
         ----------
         space : `LinearSpace`
             Space of elements which the operator is acting on.
-        a, b : ``space.field`` elements
-            Scalars to multiply ``x[0]`` and ``x[1]`` with, respectively.
+        coeffs : ``space.field`` elements
```
Comment (kohr-h) [inline odl/operator/default_ops.py]: We usually write this as `coeff1, ..., coeffN`

File: odl/operator/default_ops.py [hunk 2]
```diff
@@ -171,53 +171,157 @@ def __str__(self):
...
-        rn(3).element([2.0, 4.0, 6.0])
+        >>> op([[1, 2, 3],
+        ...     [2, 3, 4]],
+        ...    out=z)  # Returns z
+        rn(3).element([3.0, 5.0, 7.0])
         >>> z
-        rn(3).element([2.0, 4.0, 6.0])
+        rn(3).element([3.0, 5.0, 7.0])
+
+        Can also compute arbitrary weighted sums:
```
Comment (kohr-h) [inline odl/operator/default_ops.py]: It is possible to

File: odl/operator/default_ops.py [hunk 3]
```diff
@@ -171,53 +171,157 @@ def __str__(self):
...
+        else:
+            out.lincomb(self.coeffs[0], x[0], self.coeffs[1], x[1])
+            for i in range(2, n):
+                out.lincomb(1, out, self.coeffs[i], x[i])
+
         return out
 
+    @property
+    def adjoint(self):
+        """Return the adjoint.
```
Comment (kohr-h) [inline odl/operator/default_ops.py]: ~~Return~~

File: odl/operator/default_ops.py [hunk 4]
```diff
@@ -171,53 +171,157 @@ def __str__(self):
...
+        return '{}({!r}, {})'.format(self.__class__.__name__,
+                                     self.range, coeff_str)
 
-    def __str__(self):
-        """Return ``str(self)``."""
-        return "{}*x + {}*y".format(self.a, self.b)
+
+class ProductSpaceEmbedding(Operator):
+
+    """Operator mapping a vector to several copies of it::
```
Comment (kohr-h) [inline odl/operator/default_ops.py]: Same here, don't continue the summary line.

File: odl/operator/default_ops.py [hunk 5]
```diff
@@ -171,53 +171,157 @@ def __str__(self):
...
+
+        ProductSpaceEmbedding(spc, a, b)(x) == [a * x, b * x]
```
Comment (kohr-h) [inline odl/operator/default_ops.py]: This is just `BroadcastOperator(*[ScalingOperator(ci) for ci in coeffs])`, no need for a new operator.

File: odl/operator/default_ops.py [hunk 6]
```diff
@@ -171,53 +171,157 @@ def __str__(self):
...
+        """Return the adjoint.
+
+        Examples
+        --------
+        >>> r3 = odl.rn(3)
+        >>> op = LinCombOperator(r3, 1.0, 1.0)
+        >>> op.adjoint
+        ProductSpaceEmbedding(ProductSpace(rn(3), 2), 1.0, 1.0)
+        """
+        return ProductSpaceEmbedding(self.domain, *self.coeffs)
```
Comment (kohr-h) [inline odl/operator/default_ops.py]: Should be `self.range`.

File: odl/operator/default_ops.py [hunk 7]
```diff
@@ -171,53 +171,157 @@ def __str__(self):
...
+    @property
+    def adjoint(self):
+        """Return the adjoint.
+
+        Examples
+        --------
+        >>> r3 = odl.rn(3)
+        >>> op = LinCombOperator(r3, 1.0, 1.0)
+        >>> op.adjoint
+        ProductSpaceEmbedding(ProductSpace(rn(3), 2), 1.0, 1.0)
```
Comment (kohr-h) [inline odl/operator/default_ops.py]: Wrong result

**Comment by kohr-h** [comment]
> Good initiative, but too much code :-P

By that I mean that the adjoint doesn't need a new operator implementation.

File: odl/operator/default_ops.py (hunk 5 above)
Comment (adler-j) [reply odl/operator/default_ops.py]: Well the other fella is `ReductionOperator(*[ScalingOperator(ci) for ci in coeffs])` not convinced we should remove only one of them.

Comment (adler-j) [reply odl/operator/default_ops.py]: But now that you mention it there is a very clear "duality" going on here. Let's think about it for a while

Comment (kohr-h) [reply odl/operator/default_ops.py]: You're right. Maybe just make it a function instead?

Comment (adler-j) [reply odl/operator/default_ops.py]: Yes that may actually be the best option. It would also preserve `op.adjoint.adjoint == op` in some sense

**Comment by kohr-h** [comment]
Status?

**Comment by adler-j** [comment]
Basically this can be closed, If we do this, it would be according to the discussion in the comments which would be orthogonal to what has been done here.

**Pull request closed without merging**